### PR TITLE
Refactor integer range handling in the usefulness algorithm

### PR DESCRIPTION
**Pull request by Nadrieril**
Integer range handling had accumulated a lot of debt. This cleans up a lot of it.

In particular this:
- removes unnecessary conversions between `Const` and `u128`, and between `Constructor` and `IntRange`
- clearly distinguishes between on the one hand ranges of integers that may or may not be matched exhaustively, and on the other hand ranges of non-integers that are never matched exhaustively and are compared using Const-based shenanigans
- cleans up some overly complicated code paths
- generally tries to be more idiomatic.

As a nice side-effect, I measured a 10% perf increase on `unicode_normalization`.

There's one thing that I feel remains to clean up: the [overlapping range check](https://github.com/rust-lang/rust/pull/64007), which is currently quite ad-hoc. But that is intricate enough that I'm leaving it out of this PR.

There's also one little thing I'm not sure I understand: can `try_eval_bits` fail for an integer constant value in that code ? What would that mean, and how do I construct a test case for this possibility ?

### CODE REVIEW

**Comment by Nadrieril** [comment]
Actually I figured out a test case where `try_eval_bits` fails: https://github.com/rust-lang/rust/pull/66326/commits/1084eff31cad8f0d8cf2cafde4c1bc38130a09f3 . However I still don't understand what's causing this and whether this is a feature or a bug.

File: src/librustc_mir/hair/pattern/_match.rs [hunk 1]
```diff
@@ -1438,6 +1422,13 @@ impl<'tcx> IntRange<'tcx> {
     }
 }
 
+// Ignore spans when comparing, they don't carry semantic information as they are only for lints.
```
Comment (Centril) [inline src/librustc_mir/hair/pattern/_match.rs]: Prefer a doc comment here for internal rustdocs.

File: src/librustc_mir/hair/pattern/_match.rs [hunk 2]
```diff
@@ -2114,57 +2114,34 @@ fn constructor_covered_by_range<'tcx>(
...
-                from,
-                ty::ParamEnv::empty().and(ty)
-            ))) && end)
-        }
-        Single => Ok(true),
-        _ => bug!(),
-    }
+    trace!("constructor_covered_by_range {:#?}, {:#?}, {:#?}, {}", ctor, pat_from, pat_to, ty);
+
+    let to = match compare_const_vals(tcx, ctor_to, pat_to, param_env, ty) {
```
[the remark this answers is not in the document]
Comment (Nadrieril) [reply src/librustc_mir/hair/pattern/_match.rs]: Oh great, I actually started with that but thought `bool` would be more idiomatic I like `Option` better.

File: src/librustc_mir/hair/pattern/_match.rs [hunk 3]
```diff
@@ -2270,13 +2285,25 @@ fn specialize_one_pattern<'p, 'a: 'p, 'q: 'p, 'tcx>(
                     IntRange::from_pat(cx.tcx, cx.param_env, pat),
                 ) {
                     (Some(ctor), Some(pat)) => ctor.intersection(&pat).map(|_| {
+                        // Constructor splitting should ensure that all intersections we encounter
+                        // are actually inclusions.
                         let (pat_lo, pat_hi) = pat.range.into_inner();
                         let (ctor_lo, ctor_hi) = ctor.range.into_inner();
                         assert!(pat_lo <= ctor_lo && ctor_hi <= pat_hi);
```
[the remark this answers is not in the document]
Comment (Nadrieril) [reply src/librustc_mir/hair/pattern/_match.rs]: Actually I can reuse it already but I didn't notice

File: src/librustc_mir/hair/pattern/_match.rs [hunk 4]
```diff
@@ -1744,26 +1725,29 @@ fn is_useful_specialized<'p, 'a, 'tcx>(
...
 fn pat_constructor<'tcx>(
-    cx: &mut MatchCheckCtxt<'_, 'tcx>,
+    tcx: TyCtxt<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
     pat: &Pat<'tcx>,
 ) -> Option<Constructor<'tcx>> {
     match *pat.kind {
-        PatKind::AscribeUserType { ref subpattern, .. } => pat_constructor(cx, subpattern),
+        PatKind::AscribeUserType { ref subpattern, .. } => {
+            pat_constructor(tcx, param_env, subpattern)
```
[the remark this answers is not in the document]
Comment (Nadrieril) [reply src/librustc_mir/hair/pattern/_match.rs]: We use similar recursion in `specialize_one_pattern` already, which gets called repeatedly on the entries of the matrix. So I don't think this should be a problem

**Comment by Nadrieril** [comment]
Hmm, it turns out I unknowingly fixed a bug:
```rust
const BAR: &i32 = &42;
match &0 {
    BAR => {}
    _ => {}
}
```
used to flag the first branch as unreachable. This looks similar to https://github.com/rust-lang/rust/issues/65413.

**Comment by varkor** [comment]
Does this fix https://github.com/rust-lang/rust/issues/53708 as well?

**Comment by Nadrieril** [comment]
@varkor No, this was a bug specific to integer ranges and values, in the case where `try_eval_bits` fails. There is a link with #53708 however: in both cases the problem comes from consts behind references.

File: src/librustc_mir/hair/pattern/_match.rs [hunk 5]
```diff
@@ -1279,57 +1242,25 @@ impl<'tcx> IntRange<'tcx> {
...
             let (lo, hi) = (lo ^ bias, hi ^ bias);
-            // Make sure the interval is well-formed.
-            if lo > hi || lo == hi && *end == RangeEnd::Excluded {
-                None
-            } else {
-                let offset = (*end == RangeEnd::Excluded) as u128;
-                Some(IntRange { range: lo..=(hi - offset), ty, span })
+            let offset = (*end == RangeEnd::Excluded) as u128;
+            if lo > hi || (lo == hi && *end == RangeEnd::Excluded) {
+                // This hould have been caught earlier by E0030
```
Comment (varkor) [inline src/librustc_mir/hair/pattern/_match.rs]:
```suggestion
                // This should have been caught earlier by E0030.
```

File: src/librustc_mir/hair/pattern/_match.rs [hunk 6]
```diff
@@ -575,44 +575,25 @@ impl<'a, 'tcx> MatchCheckCtxt<'a, 'tcx> {
...
     /// Enum variants.
     Variant(DefId),
     /// Literal values.
-    ConstantValue(&'tcx ty::Const<'tcx>, Span),
-    /// Ranges of literal values (`2..=5` and `2..5`).
-    ConstantRange(u128, u128, Ty<'tcx>, RangeEnd, Span),
+    ConstantValue(&'tcx ty::Const<'tcx>),
+    /// Ranges of integer literal values (`2`, `2..=5` or `2..5`).
+    IntRange(IntRange<'tcx>),
+    /// Ranges of non-integer literal values (`2.0..=5.2`).
```
Comment (varkor) [inline src/librustc_mir/hair/pattern/_match.rs]: This is specifically floating-point ranges now, as far as I'm aware, so maybe we should call this `FloatRange` instead.

File: src/librustc_mir/hair/pattern/_match.rs [hunk 7]
```diff
@@ -744,37 +697,36 @@ impl<'tcx> Constructor<'tcx> {
...
+                        }
                     }
                 }
 
-                // If a constructor has not been matched, then it is missing.
-                // We add `remaining_ctors` instead of `self`, because then we can
-                // provide more detailed error information about precisely which
-                // ranges have been omitted.
-                remaining_ctors
+                // Convert the ranges back into constructors
```
Comment (varkor) [inline src/librustc_mir/hair/pattern/_match.rs]:
```suggestion
                // Convert the ranges back into constructors.
```

File: src/librustc_mir/hair/pattern/_match.rs [hunk 8]
```diff
@@ -925,12 +881,11 @@ impl<'tcx> Constructor<'tcx> {
...
             _ => match *self {
-                ConstantValue(value, _) => PatKind::Constant { value },
-                ConstantRange(lo, hi, ty, end, _) => PatKind::Range(PatRange {
-                    lo: ty::Const::from_bits(cx.tcx, lo, ty::ParamEnv::empty().and(ty)),
-                    hi: ty::Const::from_bits(cx.tcx, hi, ty::ParamEnv::empty().and(ty)),
-                    end,
-                }),
+                ConstantValue(value) => PatKind::Constant { value },
+                ConstantRange(lo, hi, end) => PatKind::Range(PatRange { lo, hi, end }),
+                IntRange(ref range) => {
```
Comment (varkor) [inline src/librustc_mir/hair/pattern/_match.rs]: This arm be collapsed on to one line.

**Review by varkor — COMMENTED** [review]
I like what you've done here. Just have a couple of very minor comments :)

**Comment by Nadrieril** [comment]
Rebased onto master and applied latest reviews :)

**Pull request merged**